### PR TITLE
Interface for writing exported API

### PR DESCRIPTION
**Pull request by jonasfj**
This is a proposal for an interface for writing the exported API.

The idea is to put everything related to managing the actual objects of the bucket into a single abstraction.

This interface isn't responsible for figuring out:
 * what to write, or,
 * when to write it.

Only **where** to write things.

The hope is that this makes it less likely that we accidentally writes file to the wrong location, forget to update them, or forget to GC particular files.

### CODE REVIEW

File: app/lib/shared/utils.dart [hunk 1]
```diff
@@ -299,4 +299,25 @@ extension StringExt on String {
...
+  /// If [suffix] is not present, it will not be removed.
+  ///
+  /// **Example:**
+  /// ```
+  /// assert('dart:io'.without(prefix: 'dart:') == 'io');
+  /// assert('file.txt'.without(suffix: '.txt') == 'file');
+  /// ```
+  String without({String prefix = '', String suffix = ''}) {
+    var val = this;
+    if (startsWith(prefix)) {
```
Comment (isoos) [inline app/lib/shared/utils.dart:315]:
```suggestion
    if (prefix.isNotEmpty && startsWith(prefix)) {
```

File: app/lib/shared/utils.dart [hunk 2]
```diff
@@ -299,4 +299,25 @@ extension StringExt on String {
...
+      val = val.substring(prefix.length);
+    }
+    if (endsWith(suffix)) {
```
Comment (isoos) [inline app/lib/shared/utils.dart:318]:
```suggestion
    if (suffix.isNotEmpty && endsWith(suffix)) {
```

File: app/lib/package/api_export/exported_api.dart [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+    }));
+  }
+
+  /// Copy binary blob from [absoluteObjectName] to this file.
+  ///
+  /// Notice that [absoluteObjectName] must be an a GCS URI including `gs://`.
+  /// This means that it must include bucket name.
+  /// Such URIs can be created with [Bucket.absoluteObjectName].
+  Future<void> copyFrom(String absoluteObjectName) async {
+    await Future.wait(_owner._prefixes.map((prefix) async {
```
Comment (isoos) [inline app/lib/package/api_export/exported_api.dart:317]: Is it possible that we have a race condition here which may update the second object earlier? (Is that bad?) -- actually this applies for most of the similar Future.wait calls in this file...

**Review by isoos — APPROVED** [review]

**Review by sigurdm — COMMENTED** [review]
Can we find a better name for this concept? Perhaps `baked_api_responses`

**Pull request merged**